### PR TITLE
Add ReadinessEndpoints to Proxy API

### PR DESCRIPTION
**Pull request by danehans**
Updates Proxy spec with `readinessEndpoints`, a list of endpoints used to verify proxy readiness before updating status fields.

Jira: [NE-206](https://jira.coreos.com/browse/NE-206)

@derekwaynecarr @mrunalp @bparees PTAL

### CODE REVIEW

File: config/v1/types_proxy.go [hunk 1]
```diff
@@ -34,6 +34,10 @@ type ProxySpec struct {
 	// Empty means unset and will not result in an env var.
 	// +optional
 	NoProxy string `json:"noProxy,omitempty"`
+
+	// readinessEndpoints is a list of endpoints used to verify readiness of the proxy.
+	// +optional
+	ReadinessEndpoints *[]string `json:"readinessEndpoints,omitempty"`
```
Comment (bparees) [inline config/v1/types_proxy.go]: i don't think we typically use pointers to slices in our api.  just make it a slice.

**Comment by bparees** [comment]
if we ignore the proxy.spec value because it failed validation (due to the health check or any other reason) how does the admin become aware that we rejected their proxy config, other than by looking at "status" and seeing that their desired spec values are not reflected there?

And how do they know why their spec was rejected?

Do we need some additional status fields for error reporting?  or is there some other mechanism by which we'll be exposing this information?

File: config/v1/types_proxy.go (hunk 1 above)
Comment (danehans) [reply config/v1/types_proxy.go]: @bparees so I understand correctly, the installer will then need to create an empty slice, correct?

**Comment by danehans** [comment]
> how does the admin become aware that we rejected their proxy config

@bparees how about adding status conditions?

File: config/v1/types_proxy.go (hunk 1 above)
Comment (bparees) [reply config/v1/types_proxy.go]: no, it can just leave the field missing.

**Comment by bparees** [comment]
> @bparees how about adding status conditions?

I think it's a reasonable first step (we may want to do more to get those conditions in front of the admin when there are issues, but having them on the resource is a necessary/appropriate starting point imho)

**Comment by danehans** [comment]
@bparees I'm glad that you feel status conditions are a good initial step. I have updated the PR accordingly, PTAL.

**Comment by bparees** [comment]
> @bparees I'm glad that you feel status conditions are a good initial step. I have updated the PR accordingly, PTAL.

sounds like things went a different direction on the slack thread (having the admission controller do all the checks up front and completely reject the spec update if it fails), so i guess we won't need status if that is the approach.  Sorry for leading you astray.

**Comment by danehans** [comment]
@bparees I removed status conditions.

File: config/v1/types_proxy.go [hunk 2]
```diff
@@ -34,6 +34,10 @@ type ProxySpec struct {
 	// Empty means unset and will not result in an env var.
 	// +optional
 	NoProxy string `json:"noProxy,omitempty"`
+
+	// readinessEndpoints is a list of endpoints used to verify readiness of the proxy.
+	// +optional
+	ReadinessEndpoints []string `json:"readinessEndpoints,omitempty"`
```
Comment (bparees) [inline config/v1/types_proxy.go]: one (minorish) concern i have w/ this is that this struct is already used as part of other config objects(embedded in them) which will not currently respect the readinessendpoints field.  So this godoc may need some clarification.  (or some apis need to be refactored such that existing embedders of the ProxySpec struct do not inherit this new field.)

/cc @adambkaplan

Comment (adambkaplan) [reply config/v1/types_proxy.go]: > or some apis need to be refactored such that existing embedders of the ProxySpec struct do not inherit this new field.

@bparees I think refactoring is the most appropriate thing to do here.

Comment (danehans) [reply config/v1/types_proxy.go]: @bparees it sounds like `BuildDefaults` is the only embedder of `ProxySpec`. @adambkaplan will create a new struct (i.e. `ProxyInfo`) based on `ProxySpec` w/o the unneeded fields and  refactor [BuildDefaults](https://github.com/openshift/api/blob/master/config/v1/types_build.go#L43-L50) to use it.

Comment (bparees) [reply config/v1/types_proxy.go]: if it works for @adambkaplan it works for me.

Comment (adambkaplan) [reply config/v1/types_proxy.go]: @bparees I discussed with @danehans on slack - this works.

File: config/v1/types_proxy.go (hunk 1 above)
Comment (danehans) [reply config/v1/types_proxy.go]: @bparees I updated `ReadinessEndpoints` to be a string.

**Pull request merged**